README.md
# ssrf_filter [![Gem](https://img.shields.io/gem/v/ssrf_filter.svg)](https://rubygems.org/gems/ssrf_filter) [![Tests](https://github.com/arkadiyt/ssrf_filter/actions/workflows/build-test.yml/badge.svg)](https://github.com/arkadiyt/ssrf_filter/actions/workflows/build-test.yml/badge.svg) [![Coverage Status](https://coveralls.io/repos/github/arkadiyt/ssrf_filter/badge.svg?branch=main)](https://coveralls.io/github/arkadiyt/ssrf_filter?branch=main) [![Downloads](https://img.shields.io/gem/dt/ssrf_filter?style=flat-square)](https://rubygems.org/gems/ssrf_filter) [![License](https://img.shields.io/github/license/arkadiyt/ssrf_filter.svg)](https://github.com/arkadiyt/ssrf_filter/blob/master/LICENSE.md)

## Table of Contents
- [What's it for](https://github.com/arkadiyt/ssrf_filter#whats-it-for)
- [Quick start](https://github.com/arkadiyt/ssrf_filter#quick-start)
- [API Reference](https://github.com/arkadiyt/ssrf_filter#api-reference)
- [Changelog](https://github.com/arkadiyt/ssrf_filter#changelog)
- [Contributing](https://github.com/arkadiyt/ssrf_filter#contributing)

### What's it for

ssrf_filter makes it easy to defend against server side request forgery (SSRF) attacks. SSRF vulnerabilities happen when you accept URLs as user input and fetch them on your server (for instance, when a user enters a link into a Twitter/Facebook status update and a content preview is generated).

Users can pass in URLs or IPs such that your server will make requests to the internal network. For example if you're hosted on AWS they can request the [instance metadata endpoint](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-metadata.html) `http://169.254.169.254/latest/meta-data/` and get your IAM credentials.

Attempts to guard against this are often implemented incorrectly, by blocking all ip addresses, not handling IPv6 or http redirects correctly, or having TOCTTOU bugs and other issues.

This gem provides a safe and easy way to fetch content from user-submitted urls. It:
- handles URIs/IPv4/IPv6, redirects, DNS, etc, correctly
- has 0 runtime dependencies
- has a comprehensive test suite (100% code coverage)
- is tested against ruby `2.6`, `2.7`, `3.0`, `3.1`, and `ruby-head`

### Quick start

1) Add the gem to your Gemfile:

```ruby
gem 'ssrf_filter', '~> 1.1.2'
```

2) In your code:

```ruby
require 'ssrf_filter'
response = SsrfFilter.get(params[:url]) # throws an exception for unsafe fetches
response.code
=> "200"
response.body
=> "<!doctype html>\n<html>\n<head>\n..."
```

### API reference

`SsrfFilter.get/.put/.post/.delete/.head/.patch(url, options = {}, &block)`

Fetches the requested url using a get/put/post/delete/head/patch request, respectively.

Params:
- `url` — the url to fetch.
- `options` — options hash (described below).
- `block` — a block that will receive the [HTTPRequest](https://ruby-doc.org/stdlib-2.4.1/libdoc/net/http/rdoc/Net/HTTPGenericRequest.html) object before it's sent, if you need to do any pre-processing on it (see examples below).

Options hash:
- `:scheme_whitelist` — an array of schemes to allow. Defaults to `%w[http https]`.
- `:resolver` — a proc that receives a hostname string and returns an array of [IPAddr](https://ruby-doc.org/stdlib-2.4.1/libdoc/ipaddr/rdoc/IPAddr.html) objects. Defaults to resolving with Ruby's [Resolv](https://ruby-doc.org/stdlib-2.4.1/libdoc/resolv/rdoc/Resolv.html). See examples below for a custom resolver.
- `:max_redirects` — Maximum number of redirects to follow. Defaults to 10.
- `:params` — Hash of params to send with the request.
- `:headers` — Hash of headers to send with the request.
- `:body` — Body to send with the request.
- `:http_options` – Options to pass to [Net::HTTP.start](https://ruby-doc.org/stdlib-2.6.4/libdoc/net/http/rdoc/Net/HTTP.html#method-c-start). Use this to set custom timeouts or SSL options.
- `:request_proc` - a proc that receives the request object, for custom modifications before sending the request.

Returns:

An [HTTPResponse](https://ruby-doc.org/stdlib-2.4.1/libdoc/net/http/rdoc/Net/HTTPResponse.html) object if the url was fetched safely, or throws an exception if it was unsafe. All exceptions inherit from `SsrfFilter::Error`.

Examples:

```ruby
# GET www.example.com
SsrfFilter.get('https://www.example.com')

# Pass params - these are equivalent
SsrfFilter.get('https://www.example.com?param=value')
SsrfFilter.get('https://www.example.com', params: {'param' => 'value'})

# POST, send custom header, and don't follow redirects
begin
  SsrfFilter.post('https://www.example.com', max_redirects: 0,
    headers: {'content-type' => 'application/json'})
rescue SsrfFilter::Error => e
  # Got an unsafe url
end

# Custom DNS resolution and request processing
resolver = proc do |hostname|
  [IPAddr.new('2001:500:8f::53')] # Static resolver
end
# Do some extra processing on the request
request_proc = proc do |request|
  request['content-type'] = 'application/json'
  request.basic_auth('username', 'password')
end
SsrfFilter.get('https://www.example.com', resolver: resolver, request_proc: request_proc)

# Stream response
SsrfFilter.get('https://www.example.com') do |response|
  response.read_body do |chunk|
    puts chunk
  end
end
```

### Changelog

Please see [CHANGELOG.md](https://github.com/arkadiyt/ssrf_filter/blob/master/CHANGELOG.md). This project follows [semantic versioning](https://semver.org/).

### Contributing

Please see [CONTRIBUTING.md](https://github.com/arkadiyt/ssrf_filter/blob/master/CONTRIBUTING.md).


CHANGELOG.md
### 1.1.2 (8/8/2023)
* Uses `Addressable::URI` to parse URLs instead of `URI` ([heri](https://github.com/kluein/ssrf_filter/pull/2))

### 1.1.1 (8/31/2022)
* Fix network connection errors if you were making https requests while using [net-http](https://github.com/ruby/net-http) 2.2 or higher ([arkadiyt](https://github.com/arkadiyt/ssrf_filter/pull/54))

### 1.1.0 (8/28/2022)
* Add support for chunked responses ([mrhaddad](https://github.com/arkadiyt/ssrf_filter/pull/30))

### 1.0.8 (8/3/2022)
* Add support for HEAD requests ([jakeyheath](https://github.com/arkadiyt/ssrf_filter/pull/38))

### 1.0.7 (10/21/2019)
* Allow passing custom options to Net::HTTP.start ([groe](https://github.com/arkadiyt/ssrf_filter/pull/26))

### 1.0.6 (2/24/2018)
* Backport a fix for a [bug](https://bugs.ruby-lang.org/issues/10054) in Ruby's http library

### 1.0.5 (1/17/2018)
* Don't send the port number in the Host header if it's HTTPS and on port 443

### 1.0.4 (1/17/2018)
* Handle relative redirects

### 1.0.3 (12/4/2017)
* Use `frozen_string_literal` pragma in all ruby files
* Handle new ruby 2.5 behavior when encountering newlines in header names

### 1.0.2 (8/3/2017)
* Block newlines and carriage returns in header names/values

### 1.0.1 (7/26/2017)
* Fixed a bug in how ipv4-compatible and ipv4-mapped addresses were handled
* Fixed a bug where the Host header did not include the port number

### 1.0.0 (7/24/2017)
* Initial release


lib/ssrf_filter/ssrf_filter.rb
# frozen_string_literal: true

require 'ipaddr'
require 'net/http'
require 'resolv'
require 'uri'

class SsrfFilter
  def self.prefixlen_from_ipaddr(ipaddr)
    mask_addr = ipaddr.instance_variable_get('@mask_addr')
    raise ArgumentError, 'Invalid mask' if mask_addr.zero?

    while (mask_addr & 0x1).zero?
      mask_addr >>= 1
    end

    length = 0
    while mask_addr & 0x1 == 0x1
      length += 1
      mask_addr >>= 1
    end

    length
  end
  private_class_method :prefixlen_from_ipaddr

  # https://en.wikipedia.org/wiki/Reserved_IP_addresses
  IPV4_BLACKLIST = [
    ::IPAddr.new('0.0.0.0/8'), # Current network (only valid as source address)
    ::IPAddr.new('10.0.0.0/8'), # Private network
    ::IPAddr.new('100.64.0.0/10'), # Shared Address Space
    ::IPAddr.new('127.0.0.0/8'), # Loopback
    ::IPAddr.new('169.254.0.0/16'), # Link-local
    ::IPAddr.new('172.16.0.0/12'), # Private network
    ::IPAddr.new('192.0.0.0/24'), # IETF Protocol Assignments
    ::IPAddr.new('192.0.2.0/24'), # TEST-NET-1, documentation and examples
    ::IPAddr.new('192.88.99.0/24'), # IPv6 to IPv4 relay (includes 2002::/16)
    ::IPAddr.new('192.168.0.0/16'), # Private network
    ::IPAddr.new('198.18.0.0/15'), # Network benchmark tests
    ::IPAddr.new('198.51.100.0/24'), # TEST-NET-2, documentation and examples
    ::IPAddr.new('203.0.113.0/24'), # TEST-NET-3, documentation and examples
    ::IPAddr.new('224.0.0.0/4'), # IP multicast (former Class D network)
    ::IPAddr.new('240.0.0.0/4'), # Reserved (former Class E network)
    ::IPAddr.new('255.255.255.255') # Broadcast
  ].freeze

  IPV6_BLACKLIST = ([
    ::IPAddr.new('::1/128'), # Loopback
    ::IPAddr.new('64:ff9b::/96'), # IPv4/IPv6 translation (RFC 6052)
    ::IPAddr.new('100::/64'), # Discard prefix (RFC 6666)
    ::IPAddr.new('2001::/32'), # Teredo tunneling
    ::IPAddr.new('2001:10::/28'), # Deprecated (previously ORCHID)
    ::IPAddr.new('2001:20::/28'), # ORCHIDv2
    ::IPAddr.new('2001:db8::/32'), # Addresses used in documentation and example source code
    ::IPAddr.new('2002::/16'), # 6to4
    ::IPAddr.new('fc00::/7'), # Unique local address
    ::IPAddr.new('fe80::/10'), # Link-local address
    ::IPAddr.new('ff00::/8') # Multicast
  ] + IPV4_BLACKLIST.flat_map do |ipaddr|
    prefixlen = prefixlen_from_ipaddr(ipaddr)

    # Don't call ipaddr.ipv4_compat because it prints out a deprecation warning on ruby 2.5+
    ipv4_compatible = IPAddr.new(ipaddr.to_i, Socket::AF_INET6).mask(96 + prefixlen)
    ipv4_mapped = ipaddr.ipv4_mapped.mask(80 + prefixlen)

    [ipv4_compatible, ipv4_mapped]
  end).freeze

  DEFAULT_SCHEME_WHITELIST = %w[http https].freeze

  DEFAULT_RESOLVER = proc do |hostname|
    ::Resolv.getaddresses(hostname).map { |ip| ::IPAddr.new(ip) }
  end

  DEFAULT_MAX_REDIRECTS = 10

  VERB_MAP = {
    get: ::Net::HTTP::Get,
    put: ::Net::HTTP::Put,
    post: ::Net::HTTP::Post,
    delete: ::Net::HTTP::Delete,
    head: ::Net::HTTP::Head,
    patch: ::Net::HTTP::Patch
  }.freeze

  FIBER_HOSTNAME_KEY = :__ssrf_filter_hostname
  FIBER_ADDRESS_KEY = :__ssrf_filter_address

  class Error < ::StandardError
  end

  class InvalidUriScheme < Error
  end

  class PrivateIPAddress < Error
  end

  class UnresolvedHostname < Error
  end

  class TooManyRedirects < Error
  end

  class CRLFInjection < Error
  end

  %i[get put post delete head patch].each do |method|
    define_singleton_method(method) do |url, options = {}, &block|
      ::SsrfFilter::Patch::SSLSocket.apply!
      ::SsrfFilter::Patch::Resolv.apply!

      original_url = url
      max_redirects = options[:max_redirects] || DEFAULT_MAX_REDIRECTS
      url = url.to_s

      (max_redirects + 1).times do
        uri = URI(url)

        public_addresses = get_public_ip_addresses(uri, resolver: options[:resolver],
                                                        scheme_whitelist: options[:scheme_whitelist],
                                                        raise_expection: true)

        raise PrivateIPAddress, "Hostname '#{uri.hostname}' has no public ip addresses" if public_addresses.empty?

        response, url = fetch_once(uri, public_addresses.sample.to_s, method, options, &block)
        return response if url.nil?
      end

      raise TooManyRedirects, "Got #{max_redirects} redirects fetching #{original_url}"
    end
  end

  def self.unsafe_url?(url)
    uri = Addressable::URI.parse(url)

    get_public_ip_addresses(uri).empty?
  end

  def self.get_public_ip_addresses(uri, resolver: nil, scheme_whitelist: nil, raise_expection: false)
    scheme_whitelist ||= DEFAULT_SCHEME_WHITELIST
    resolver ||= DEFAULT_RESOLVER

    unless scheme_whitelist.include?(uri.scheme)
      raise InvalidUriScheme, "URI scheme '#{uri.scheme}' not in whitelist: #{scheme_whitelist}" if raise_expection

      return []
    end

    hostname = uri.hostname
    ip_addresses = resolver.call(hostname)

    if ip_addresses.empty?
      raise UnresolvedHostname, "Could not resolve hostname '#{hostname}'" if raise_expection

      return []
    end

    ip_addresses.reject(&method(:unsafe_ip_address?))
  end
  private_class_method :get_public_ip_addresses

  def self.unsafe_ip_address?(ip_address)
    return true if ipaddr_has_mask?(ip_address)

    return IPV4_BLACKLIST.any? { |range| range.include?(ip_address) } if ip_address.ipv4?
    return IPV6_BLACKLIST.any? { |range| range.include?(ip_address) } if ip_address.ipv6?

    true
  end
  private_class_method :unsafe_ip_address?

  def self.ipaddr_has_mask?(ipaddr)
    return true unless ipaddr

    range = ipaddr.to_range
    range.first != range.last
  end
  private_class_method :ipaddr_has_mask?

  def self.host_header(hostname, uri)
    # Attach port for non-default as per RFC2616
    if (uri.port == 80 && uri.scheme == 'http') ||
       (uri.port == 443 && uri.scheme == 'https')
      hostname
    else
      "#{hostname}:#{uri.port}"
    end
  end
  private_class_method :host_header

  def self.fetch_once(uri, ip, verb, options, &block)
    if options[:params]
      params = uri.query ? ::URI.decode_www_form(uri.query).to_h : {}
      params.merge!(options[:params])
      uri.query = ::URI.encode_www_form(params)
    end

    hostname = uri.hostname
    uri.hostname = ip

    request = VERB_MAP[verb].new(uri)
    request['host'] = host_header(hostname, uri)

    Array(options[:headers]).each do |header, value|
      request[header] = value
    end

    request.body = options[:body] if options[:body]

    options[:request_proc].call(request) if options[:request_proc].respond_to?(:call)
    validate_request(request)

    http_options = options[:http_options] || {}
    http_options[:use_ssl] = (uri.scheme == 'https')

    with_forced_hostname(hostname, ip) do
      ::Net::HTTP.start(uri.hostname, uri.port, **http_options) do |http|
        http.request(request) do |response|
          case response
          when ::Net::HTTPRedirection
            url = response['location']
            # Handle relative redirects
            url = "#{uri.scheme}://#{hostname}:#{uri.port}#{url}" if url.start_with?('/')
            return nil, url
          else
            block&.call(response)
            return response, nil
          end
        end
      end
    end
  end
  private_class_method :fetch_once

  def self.validate_request(request)
    # RFC822 allows multiline "folded" headers:
    # https://tools.ietf.org/html/rfc822#section-3.1
    # In practice if any user input is ever supplied as a header key/value, they'll get
    # arbitrary header injection and possibly connect to a different host, so we block it
    request.each do |header, value|
      if header.count("\r\n") != 0 || value.count("\r\n") != 0
        raise CRLFInjection, "CRLF injection in header #{header} with value #{value}"
      end
    end
  end
  private_class_method :validate_request

  def self.with_forced_hostname(hostname, ip, &_block)
    ::Thread.current[FIBER_HOSTNAME_KEY] = hostname
    ::Thread.current[FIBER_ADDRESS_KEY] = ip
    yield
  ensure
    ::Thread.current[FIBER_HOSTNAME_KEY] = nil
    ::Thread.current[FIBER_ADDRESS_KEY] = nil
  end
  private_class_method :with_forced_hostname
end
